### DISCUSSION TITLE
How to judge libcurl No connections available

### DISCUSSION THREAD

**Discussion by wzssga** [curl/curl#10758] [Q&A] [answered]
Aim ：
I hope to be able to control the number of requests for libcurl http2, so I use CURLINFO_NUM_CONNECTS, 
I want to reduce the number of concurrent download tasks in the upper layer when libcurl “CURLE_NO_CONNECTION_AVAILABLE”。

Now I can get  ‘No connections available’ by OnDebugCallBack Info：
![image](https://user-images.githubusercontent.com/33863256/224886100-c2336181-083e-496c-91aa-f41b0796e3c1.png)

Question：
Can I get the status of the request through other methods?, such as easy_get_info: data->mstate(pending), but I found that curl_easy_getinfo does not provide this parameter
![image](https://user-images.githubusercontent.com/33863256/224886674-eb10750b-e80f-4fb6-be67-dcb5ca46d9b9.png)

**Comment by bagder** [comment, marked as answer]
No, there is no API way currently to get to know the internal state for an easy handle. Like for example when it is set to `PENDING` state because it is prohibited to start a transfer.

I would be willing to discuss ways to expose this.

**Reply by wzssga** [reply]
Thanks for your answer.

Maybe can add new key in the getinfo API to get easy_handle's  mstate.

![image](https://user-images.githubusercontent.com/33863256/225279884-a8eede52-225c-4351-b798-e34da2cb008d.png)

**Reply by bagder** [reply]
Maybe, but:

1. I would not like to expose the internal state just like that. It would need a dedicated set of externally defined states so that we can change internals in a future without breaking the public API.
2. Providing this info is very "state of the moment" as it might change again in the next call.

**Reply by wzssga** [reply]
> *1. It would need a dedicated set of externally defined states so that we can change internals in a future without breaking the public API.。

Is it possible to add "CURLINFO_HANDLE_PENDED" param to determine whether easy_hadle is currently in the pending state, so that there is no need to expose all internal states.
![image](https://user-images.githubusercontent.com/33863256/225308402-245d421c-6f1d-4959-9d8e-659c97e55f14.png)


> * 2.Providing this info is very "state of the moment" as it might change again in the next call.。

After each 'curl_multi_perform', I will get the pending status of the newly added easy_handle through 'curl_easy_getinfo(handle,CURLINFO_HANDLE_PENDED,bIsPend)' to judge whether there are too many concurrent requests.

Is this solution feasible？

**Reply by bagder** [reply]
Feasible, sure. I don't think it's very convenient though.
